File: csharp/zipper/Zipper.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class BinTree
{
    public BinTree(int value, BinTree left, BinTree right) => (Value, Left, Right) = (value, left, right);
    public int Value { get; }
    public BinTree Left { get; }
    public BinTree Right { get; }
    public override bool Equals(object obj)
    {
        if (obj is not BinTree b || b.Value != Value) return false;
        bool leftIsSame = (b.Left == null && Left == null) || (Left != null && Left.Equals(b.Left));
        bool rightIsSame = (b.Right == null && Right == null) || (Right != null && Right.Equals(b.Left));
        return leftIsSame && rightIsSame;
    }
    public override int GetHashCode() => base.GetHashCode();
}

public class Zipper
{   
    public Zipper(IEnumerable<(BinTree, bool)> parents, BinTree focus)
    {
        Parents = parents;
        Focus = focus;
    }
    public IEnumerable<(BinTree, bool)> Parents { get; }
    public BinTree Focus { get; }
    public int Value()
    {
        return Focus.Value;
    }

    public Zipper SetValue(int newValue)
    {
        return FromTree(new BinTree(newValue, Focus.Left, Focus.Right), Parents);
    }

    public Zipper SetLeft(BinTree binTree)
    {
        return FromTree(new BinTree(Focus.Value, binTree, Focus.Right), Parents);
    }

    public Zipper SetRight(BinTree binTree) 
    {
        return FromTree(new BinTree(Focus.Value, Focus.Left, binTree), Parents);
    }

    public Zipper Left()
    {
        return Focus.Left != null ? FromTree(Focus.Left, Parents.Prepend((Focus, true))) : null;
    }

    public Zipper Right()
    {
        return Focus.Right != null ? FromTree(Focus.Right, Parents.Prepend((Focus, false))) : null;
    }

    public Zipper Up()
    {
        if (!Parents.Any()) return null;
        var (tree, isLeft) = Parents.First();
        var rest = Parents.Skip(1);
        var newFocus = new BinTree(tree.Value, isLeft ? Focus : tree.Left, isLeft ? tree.Right : Focus);
        return FromTree(newFocus, rest);
    }

    public BinTree ToTree()
    {
        if (Parents.Count() == 0) return Focus;
        Zipper actual = this;
        while (actual.Parents.Count() > 0)
        {
            actual = actual.Up();
        }
        return actual.Focus;
    }

    public static Zipper FromTree(BinTree tree, IEnumerable<(BinTree, bool)> parents = null)
    {
        return new Zipper(parents ?? Enumerable.Empty<(BinTree, bool)>(), tree);
    }
    public override bool Equals(object obj) => (obj as Zipper).Focus.Equals(this.Focus);
    public override int GetHashCode() => base.GetHashCode();
}

File: elixir/zipper/lib/zipper.ex
defmodule Zipper do

  import BinTree

  @type t :: %Zipper{focus: BinTree.t(), parents: [{BinTree.t(), boolean()}]}
  defstruct [:focus, :parents]

  @doc """
  Get a zipper focused on the root node.
  """
  @spec from_tree(BinTree.t(), [{BinTree.t(), boolean()}]) :: Zipper.t()
  def from_tree(bin_tree, new_parents \\ []) do
    %Zipper{focus: bin_tree, parents: new_parents}
  end

  @doc """
  Get the complete tree from a zipper.
  """
  @spec to_tree(Zipper.t()) :: BinTree.t()
  def to_tree(zipper) do
    case zipper.parents do
      [] -> zipper.focus
      _  -> up(zipper) |> to_tree()
    end
  end

  @doc """
  Get the value of the focus node.
  """
  @spec value(Zipper.t()) :: any
  def value(zipper) do
    zipper.focus.value
  end

  @doc """
  Get the left child of the focus node, if any.
  """
  @spec left(Zipper.t()) :: Zipper.t() | nil
  def left(zipper) do
    if is_nil(zipper.focus.left), do: nil, else: from_tree(zipper.focus.left, [{zipper.focus, true}|zipper.parents])
  end

  @doc """
  Get the right child of the focus node, if any.
  """
  @spec right(Zipper.t()) :: Zipper.t() | nil
  def right(zipper) do
    if is_nil(zipper.focus.right), do: nil, else: from_tree(zipper.focus.right, [{zipper.focus, false}|zipper.parents])
  end

  @doc """
  Get the parent of the focus node, if any.
  """
  @spec up(Zipper.t()) :: Zipper.t() | nil
  def up(zipper) do
    if zipper.parents |> Enum.count() == 0 do
      nil
    else
      [{tree, is_left}|rest] = zipper.parents
      new_focus = %BinTree{
        value: tree.value,
        left: if(is_left, do: zipper.focus, else: tree.left),
        right: if(is_left, do: tree.right, else: zipper.focus)
      }
      from_tree(new_focus, rest)
    end
  end

  @doc """
  Set the value of the focus node.
  """
  @spec set_value(Zipper.t(), any) :: Zipper.t()
  def set_value(zipper, value) do
    %{zipper | focus: %{zipper.focus | value: value}}
  end

  @doc """
  Replace the left child tree of the focus node.
  """
  @spec set_left(Zipper.t(), BinTree.t() | nil) :: Zipper.t()
  def set_left(zipper, left) do
    %{zipper | focus: %{zipper.focus | left: left}}
  end

  @doc """
  Replace the right child tree of the focus node.
  """
  @spec set_right(Zipper.t(), BinTree.t() | nil) :: Zipper.t()
  def set_right(zipper, right) do
    %{zipper | focus: %{zipper.focus | right: right}}
  end
end
